### DISCUSSION TITLE
How pass dynamically env value to proxy library on nuxt config

### DISCUSSION THREAD

**Discussion by viniciuspinatti** [nuxt/nuxt#26945] [Questions]
Hey guys!

I'm having some issues with a Nuxt 3 application. I'm migrating an app from Nuxt 2 to Nuxt 3 and I had some problems with websocket. But I solved it using the module @nuxt-alt/proxy. My problem is I can't use the env vars NUXT_PUBLIC_BROWSER_BASE_URL and NUXT_PUBLIC_WS_BASE_URL to set the `target` in proxy library config. They only work in runtimeConfig, and the code below doesn't work for proxy configuration when I build to prod (I got an empty value for both targets).

```
export default defineNuxtConfig({
  modules: [
    '@nuxt-alt/proxy',
    ...
  ],
  runtimeConfig: {
    internalApiKey: process.env.NUXT_INTERNAL_API_KEY,
    public: {
      browserBaseURL: process.env.NUXT_PUBLIC_BROWSER_BASE_URL,
      environment: process.env.NUXT_PUBLIC_ENVIRONMENT,
      wsBaseUrl: process.env.NUXT_PUBLIC_WS_BASE_URL,
    },
  },
  proxy: {
    debug: false,
    experimental: {
      listener: true,
    },
    proxies: {
      '/api': {
        target: process.env.NUXT_PUBLIC_BROWSER_BASE_URL,
        changeOrigin: true,
        ws: false,
      },
      '/ws': {
        target: process.env.NUXT_PUBLIC_WS_BASE_URL,
        changeOrigin: true,
        ws: true,
      },
    },
  },
  ...
})
```

But if I change the target to a literal url string (example below), then both start to work.

```
  proxy: {
    debug: false,
    experimental: {
      listener: true,
    },
    proxies: {
      '/api': {
        target: 'https://my-api-url.com',
        changeOrigin: true,
        ws: false,
      },
      '/ws': {
        target: 'wss://my-api-url.com',
        changeOrigin: true,
        ws: true,
      },
    },
  },
```

I cannot use a fixed value. I need to use environment config because I get the values from AWS SSM parameters for each environment (dev, demo, prod) in my AWS container task.
I already tried to use VITE_ vars with `import.meta.env` but it didn't work.
I also tried to create a module file to get the values from runtime config (example below).

```
// proxy.ts on modules directory
import { defineNuxtModule, installModule } from '@nuxt/kit'

export default defineNuxtModule({
  async setup(options, nuxt) {
    await installModule(
      '@nuxt-alt/proxy',
      {
        debug: false,
        experimental: {
          listener: true,
        },
        proxies: {
          '/api': {
            target: nuxt.options.runtimeConfig.public.browserBaseURL,
            changeOrigin: false,
            ws: false,
          },
          '/ws': {
            target: nuxt.options.runtimeConfig.public.wsBaseUrl,
            changeOrigin: false,
            ws: true,
          },
        },
      }
    )
  },
})
```

I didn't have success yet. Then only way that it worked was putting the hard code URL string.
Note: the values are being detected from the environment because I checked that runtimeConfig values are being populated correctly.
I'm open to any solution or idea to avoid put a fixed value there.

**Comment by jwhazel** [comment]
Where does your application get built (GitHub actions, Jenkins, CodePipeline, etc)?

**Reply by viniciuspinatti** [reply]
On Gitlab. There is a job in a pipeline that creates the AWS infra and the ECS tasks executing a AWS cloudformation json. The task has the container with the environment variables and secrets. Something like this:

```
"TaskDefinitionApp": {
    "Type": "AWS::ECS::TaskDefinition",
    "Properties": {
      "ContainerDefinitions": [
        {
          "Cpu": 1024,
          "DisableNetworking": false,
          "Environment": [
            {
              "Name": "NUXT_PUBLIC_ENVIRONMENT",
              "Value": { "Ref": "SecretsEnvironment" }
            }
          ],
          "Secrets": [
            {
              "Name": "NUXT_PUBLIC_BROWSER_BASE_URL",
              "ValueFrom": {"Fn::Sub": "arn:aws:ssm:${AWS::Region}:${AWS::AccountId}:parameter/config/app_${SecretsEnvironment}/api.base.url"}
            },
            {
              "Name": "NUXT_PUBLIC_WS_BASE_URL",
              "ValueFrom": { "Fn::Sub": "arn:aws:ssm:${AWS::Region}:${AWS::AccountId}:parameter/config/app_${SecretsEnvironment}/ws.base.url" }
            }
          ],
          "Essential": true,
          "Image": { "Ref": "MyAppImage" },
          "Memory": 2048,
          "Name": "my-app-name",
          "PortMappings": [{ "ContainerPort": 3000, "Protocol": "tcp" }]
        }
      ],
      "Cpu": "1024",
      "ExecutionRoleArn": { "Ref": "TaskExecutionRoleArn" },
      "Family": {
        "Fn::Sub": "my-app-name-task-${EnvironmentName}"
      },
      "Memory": "2048",
      "NetworkMode": "awsvpc"
    }
  }
```

**Reply by jwhazel** [reply]
It looks to me like you’re setting env variables as part of your infrastructure. Nuxt 3 consumes env variables _at build time_, not at runtime.

**Reply by viniciuspinatti** [reply]
But how then the runtimeConfig values are being populated correctly? The `browserBaseURL`, `environment` and `wsBaseUrl` are receiving the correct value and there is no .env file with values to populate them. This is what I didn't understand.

**Comment by lorenzo-kun** [comment]
@viniciuspinatti Did you manage to resolve this or find a workaround? I've found myself in the exact same situation as you - migrating from Nuxt 2 to Nuxt 3 and having problems reading the .env variables in a ECS task for web socket proxy configuration for the `@nuxt-alt/proxy` module.

**Reply by viniciuspinatti** [reply]
@lorenzo-kun I'm using a workaround to fix this. I created multiple nuxt.config.ts files with a pattern in name like `nuxt-ENV_NAME.config.ts`. For example:
```
nuxt-dev.config.ts
nuxt-prod.config.ts
...
```

Each file has the URL string to proxy hard coded.

Then in the gitlab pipeline I have a script that get the tag name environment and pass it as an arg for docker file:

build container script
```
build_image() {
    ...
    docker build \
        --build-arg DEPLOY_CONFIG_FILE="nuxt-${tag}.config.ts" \
        --file "${docker_file}" \
        --tag ${repo_url} \
        $(if [[ "${force_rebuild}" = "true" ]]; then echo -n "--no-cache --pull"; fi) \
        "${path}"
}
```

And on dockerfile I use the argument to copy the correct config file but on container the name is the default name nuxt.config.ts

dockerfile
```
...
ARG DEPLOY_CONFIG_FILE

RUN mkdir /app
WORKDIR /app
ENV PATH /app/node_modules/.bin:$PATH
COPY package.json /app/package.json
COPY package-lock.json /app/package-lock.json
RUN npm install --omit=dev
COPY . /app
COPY $DEPLOY_CONFIG_FILE /app/nuxt.config.ts
...
```

It is working now. There are two considerations about this approach:
- Have multiple config files is a bad thing because if you change something on the nuxt config, you should do the same for the all files (unfortunately it is an existing redundancy);
- The best approach maybe that I didn't test yet would be try to implement the nuxt-alt/proxy way to handle the websocket connection (there are few files inside of this library). This could avoid the use of ENV values outside `runtimeConfig`.

**Reply by lorenzo-kun** [reply]
Thanks for getting back to me @viniciuspinatti! I tried a bunch of different things too, but also came to the conclusion that I just need to build a different image for each environment.

I haven't fully tested this out in my pipelines yet, but I'm just planning to pass the proxy target URL as an arg into docker. So the docker build command would be
```shell
docker build --build-arg PROXY_TARGET_URL=${proxy_target_url} .
```

And the Dockerfile would be like this:
```dockerfile
...
ARG PROXY_TARGET_URL
ENV PROXY_TARGET_URL=${PROXY_TARGET_URL}
RUN npm run build
...
```

And I'll just have one nuxt.config.ts (which should address your first concern about having multiple config files)
```typescript
export default defineNuxtConfig({
  ...,
  proxy: {
    debug: false,
    experimental: {
      listener: true,
    },
    proxies: {
      '/socket.io': {
        target: (process.env.PROXY_TARGET_URL || 'http://localhost:1337').replace('http', 'ws'),
        ws: true,
      },
    },
  }.
  ...,
});
```

Your second point is a good idea though - I might also try to spend some time looking into other ways to configure the nuxt-alt/proxy without using ENV values 🤔

**Reply by viniciuspinatti** [reply]
Hey @lorenzo-kun that's great! I'll try to implement using the way that you did to avoid the multiple nuxt config files. If I get a new way to setup this or have any success with adapting the content of nuxt-alt/proxy package for my scenario I'll let you know. Thank you.

**Reply by cdecinkoKnight** [reply]
I am wrestling with the same issue. I don't want to hard code the value for m proxy target URL. If process.env.xxx is just a string, it should just work. But so far it fails to put the value there.

**Reply by lorenzo-kun** [reply]
@cdecinkoKnight I believe the issue is with the timing - I dug into the nuxt-alt/proxy source code a bit and the `process.env.xxx` is only pulled into the module options at build time, and doesn't look for the environment values at runtime. So the "problem" is with nuxt-alt/proxy, not nuxt itself, since there is a way for modules to use runtime configs.

@viniciuspinatti FWIW, I managed to get it working, but I essentially just copied the nuxt-alt/proxy source code into a local module and modified it to pull from the runtime config if I don't provide a target. I was considering raising a PR on the nuxt-alt/proxy repo if I could get it working more generically, but I couldn't figure it out and my modification is very specific to my requirements 😅 

In case you're curious, my nuxt.config.ts now looks like this:
```typescript
export default defineNuxtConfig({
  ...,
  'my-custom-proxy': {
    proxies: {
      '/socket.io': {
        ws: true,
      },
    },
  }.
  ...,
});
```

And in my version of the [socket-plugin.nitro.ts](https://github.com/nuxt-alt/proxy/blob/main/src/runtime/socket-plugin.nitro.ts#L32) after the options are initialized, I do this:
```typescript
Object.keys(options.proxies!).forEach(async (context) => {
  let opts = initializeOpts(options.proxies![context]);
  
  if (!opts) return;
  
  if (!opts.target?.toString!())
      opts.target = useRuntimeConfig().public.baseApiUrl; // I couldn't figure out how to dynamically specify which runtime config to use
  
  if (opts.ws)
      opts.target = opts.target?.toString!().replace(/^http(?=s*:\/\/)/, 'ws');

  // ... other configuration
});
```

It's very hacky, but it works and uses the ENV value properly at runtime 😄
